lib/workspaces/pnpm-workspaces-parser.ts

```diff
@@ -0,0 +1,164 @@
+// a project that's part of a workspace, we need to retrieve it from
+// its corresponding package.json
+function computeProjectVersionMaps(root: string, targetFiles) {
+  const projectsVersionMap = {};
+  for (const directory of Object.keys(targetFiles)) {
+    const packageJsonFileName = pathUtil.join(directory, 'package.json');
+    const packageJson = getFileContents(root, packageJsonFileName);
+
+    try {
+      const parsedPkgJson = lockFileParser.parsePkgJson(packageJson.content);
+      const projectVersion = parsedPkgJson.version;
+      projectsVersionMap[
+        normalizeFilePath(pathUtil.relative(root, directory))
+      ] = projectVersion;
+    } catch (err: any) {
+      debug(
+        `Error getting version for project: ${packageJsonFileName}. ERROR: ${err}`,
+      );
+      continue;
+    }
+  }
+  return projectsVersionMap;
+}
+
+export async function processPnpmWorkspaces(
+  root: string,
+  settings: {
+    strictOutOfSync?: boolean;
+    dev?: boolean;
+    optional?: boolean;
+  },
+  targetFiles: string[],
+): Promise<MultiProjectResultCustom> {
+  const pnpmTargetFiles = sortTargetFiles(targetFiles, PNPM_ROOT_FILES);
+
+  debug(`Processing potential Pnpm workspaces (${targetFiles.length})`);
+
+  const result: MultiProjectResultCustom = {
+    plugin: {
+      name: 'snyk-nodejs-pnpm-workspaces',
+      runtime: process.version,
+    },
+    scannedProjects: [],
+  };
+
+  let pnpmWorkspacesMap = {};
+  const pnpmWorkspacesFilesMap = {};
+
+  let rootWorkspaceManifestContent = {};
+  const projectsVersionMap = computeProjectVersionMaps(root, pnpmTargetFiles);
+
+  // the folders must be ordered highest first
+  for (const directory of Object.keys(pnpmTargetFiles)) {
+    debug(`Processing ${directory} as a potential Pnpm workspace`);
+    let isPnpmWorkspacePackage = false;
+    let isRootPackageJson = false;
+    const packageJsonFileName = pathUtil.join(directory, 'package.json');
+    const packageJson = getFileContents(root, packageJsonFileName);
+    pnpmWorkspacesMap = {
+      ...pnpmWorkspacesMap,
+      ...getWorkspacesMap(root, directory, packageJson),
+    };
+
+    for (const workspaceRoot of Object.keys(pnpmWorkspacesMap)) {
+      const match = packageJsonBelongsToWorkspace(
+        packageJsonFileName,
+        pnpmWorkspacesMap,
+        workspaceRoot,
+      );
+      if (match) {
+        debug(`${packageJsonFileName} matches an existing workspace pattern`);
+        pnpmWorkspacesFilesMap[packageJsonFileName] = {
+          root: workspaceRoot,
+        };
+        isPnpmWorkspacePackage = true;
+      }
+      if (packageJsonFileName === workspaceRoot) {
+        isRootPackageJson = true;
+        rootWorkspaceManifestContent = JSON.parse(packageJson.content);
+      }
+    }
+
+    if (!(isPnpmWorkspacePackage || isRootPackageJson)) {
+      debug(
+        `${packageJsonFileName} is not part of any detected workspace, skipping`,
+      );
+      continue;
+    }
```

Comment:
My understanding is here this would be picked up as a solo project if we were running `--all-projects`

Reply:
Yes